src/umpire/strategy/ResourceAwarePool.cpp

```diff
@@ -0,0 +1,549 @@
+// Copyright (c) 2016-24, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#include "umpire/strategy/ResourceAwarePool.hpp"
+
+#include "umpire/Allocator.hpp"
+#include "umpire/event/event.hpp"
+#include "umpire/strategy/PoolCoalesceHeuristic.hpp"
+#include "umpire/strategy/mixins/AlignedAllocation.hpp"
+#include "umpire/util/Macros.hpp"
+#include "umpire/util/memory_sanitizers.hpp"
+
+namespace umpire {
+namespace strategy {
+
+ResourceAwarePool::ResourceAwarePool(const std::string& name, int id, Allocator allocator,
+                                     const std::size_t first_minimum_pool_allocation_size,
+                                     const std::size_t next_minimum_pool_allocation_size, std::size_t alignment,
+                                     PoolCoalesceHeuristic<ResourceAwarePool> should_coalesce) noexcept
+    : AllocationStrategy{name, id, allocator.getAllocationStrategy(), "ResourceAwarePool"},
+      mixins::AlignedAllocation{alignment, allocator.getAllocationStrategy()},
+      m_should_coalesce{should_coalesce},
+      m_first_minimum_pool_allocation_size{first_minimum_pool_allocation_size},
+      m_next_minimum_pool_allocation_size{next_minimum_pool_allocation_size}
+{
+  UMPIRE_LOG(Debug, " ( "
+                        << "name=\"" << name << "\""
+                        << ", id=" << id << ", allocator=\"" << allocator.getName() << "\""
+                        << ", first_minimum_pool_allocation_size=" << m_first_minimum_pool_allocation_size
+                        << ", next_minimum_pool_allocation_size=" << m_next_minimum_pool_allocation_size
+                        << ", alignment=" << alignment << " )");
+}
+
+ResourceAwarePool::~ResourceAwarePool()
+{
+  UMPIRE_LOG(Debug, "Releasing free blocks to device");
+  m_is_destructing = true;
+  release();
+}
+
+void* ResourceAwarePool::allocate(std::size_t UMPIRE_UNUSED_ARG(bytes))
+{
+  void* ptr{nullptr};
+  UMPIRE_ERROR(
```

Comment:
Should ask users if they would like to specify a "default" resource that could be used in this case.

Reply:
For debugging purposes, I think it would be better for a user to create a helper function in their own code that returns a default resource and then always call the RAP allocate_resource with that returned resource rather than us  supplying it - but I can also put this on the discussion list for next umpire user meeting